Start using Logs.src instead of tags

aryx opened the pull request:
We're currently abusing tags for something it was not designed for.
The standard convention in the OCaml libraries world
is to use the Logs "src" to separate the logging
done by different libraries, so let's do the same
in the semgrep codebase.
See https://erratique.ch/software/logs/doc/Logs/index.html#srcs

test plan:
```
$ make check
...
[00.04][WARNING]: Support for Jsonnet rules is experimental and currently meant for internal use only. The syntax may change or be removed at any point.
[00.92][ERROR](commons.pcre): PCRE error: BadUTF8 on input "\127ELF\002\001\001\003\000\000\000\000\000\000\000\000\003\000>\000\001\000\000\0000\021\200\001\000\000\000\00
...
```

aryx committed:
Start using Logs.src instead of tags

We're currently abusing tags for something it was not designed for.
The standard convention in the OCaml libraries world
is to use the Logs "src" to separate the logging
done by different libraries, so let's do the same
in the semgrep codebase.
See https://erratique.ch/software/logs/doc/Logs/index.html#srcs

test plan:
```
$ make check
...
[00.04][WARNING]: Support for Jsonnet rules is experimental and currently meant for internal use only. The syntax may change or be removed at any point.
[00.92][ERROR](commons.pcre): PCRE error: BadUTF8 on input "\127ELF\002\001\001\003\000\000\000\000\000\000\000\000\003\000>\000\001\000\000\0000\021\200\001\000\000\000\00
...
```

aryx commented:
In my next PR (or next commit) I will provide a way to dynamically specify the addition src we want to show the logs for, like SEMGREP_LOGS_SRC="common.pcre,bos,fpath" so we show the logs of this library too.

aryx commented:
After that we can have a separate PR to remove our current use of tags and switch to Logs.src for those, which will simplify a lot Logs_.ml which got complicated with our use of tags.

aryx committed:
misc

aryx committed:
Add a new LOG_SRCS env var to control which Log src to display

also get rid of the semgrep specific stuff in Logs_.ml

aryx committed:
misc

aryx commented:
test plan:
```
$ LOG_SRCS=commons.pcre,bos  ./bin/osemgrep --experimental --config semgrep.jsonnet --error --strict --exclude tests
...
00.04][WARNING]: Support for Jsonnet rules is experimental and currently meant for internal use only. The syntax may change or be removed at any point.
[00.95][ERROR](commons.pcre): PCRE error: BadUTF8 on input "\127ELF\002\001\001\003\000\000\000\000\000\000\000\000\003\000>\000\001\000\000\0000\021\200\001\000\000\000\000@\000\000\000\000\000\000\0000\218\234\011\000\000\000\000\000\000\000\000@\0008\000\r\000@\000*\000)\000\006\000\000\000\004\000\000\000@\000\000\000\000\000\000\000@\000\000\000\000\000\000\000@\000\
...
```

aryx committed:
update .mli comment

IagoAbal commented on libs/commons/Logs_.ml:
```diff
@@ -142,35 +123,55 @@ let has_nonempty_intersection tag_str_list tag_set =
   let select_all_debug_messages = List.mem "all" require_one_of_these_tags in
-  let report _src level ~over k msgf =
+
+  (* here we go ... this is quite complicated *)
```
Did you mean to write some other comment? 😄

IagoAbal commented on libs/commons/Logs_.ml:
```diff
@@ -267,27 +243,29 @@ let read_level_from_env vars =
-      [ "PYTEST_SEMGREP_LOG_TAGS"; "SEMGREP_LOG_TAGS" ]) ~level () =
+let setup ?(highlight_setting = Std_msg.get_highlight_setting ())
+    ?log_to_file:opt_file ?(require_one_of_these_tags = default_tags)
+    ?(read_level_from_env_vars = [ "LOG_LEVEL" ])
```
Note GitLab was reporting perf problems due to all the logging that was produced by Semgrep, IIRC due to GitLab CI running with `LOG_LEVEL=debug`, Martin quick-fixed this by renaming it to `SEMGREP_LOG_LEVEL`. There are also changelog entries about these renames, so we would need to write more changelog entries for this.

aryx replied:
Logs_.ml is supposed to be independent of semgrep, but we can use SEMGREP_LOG_LEVEL in
the caller to Logs_.setup(), in CLI.ml and Core_CLI.ml

mjambon replied:
There at least 3 levels with the same name:
* Semgrep, the organization or software vendor
* the `semgrep` repo (+ `semgrep-proprietary`) = libraries provided by the Semgrep vendor + Semgrep application
* the Semgrep application

The `SEMGREP_` prefix here is for namespacing purposes and refers to one of the three things above. I'm not totally sure which.

mjambon replied:
Don't use `LOG_LEVEL` for the reason @IagoAbal mentioned. In general, environment variables are tricky because they're passed from process to process invisibly. I suggest not setting a default environment variable if the goal is to be a library that can be used outside of semgrep. Make it the application's job to explicitly enable an environment variable.

aryx replied:
I think LOG_LEVEL is fine as a default, but yes I'll update the caller to not use this variable name.

aryx committed:
fix js

aryx committed:
update snapshot

aryx committed:
remove useless comment

mjambon commented on libs/commons/Logs_.ml:
```diff
@@ -142,35 +123,54 @@ let has_nonempty_intersection tag_str_list tag_set =
-  vars |> read_from_environment_variables
-  |> Option.map (String.split_on_char ',')
+let read_csv_from_env_vars (vars : string list) : string list option =
+  vars |> read_from_env_vars |> Option.map (String.split_on_char ',')
```
CSV refers to a specific format. Please don't call this CSV because it's not how CSV is parsed.

aryx replied:
true, csv usually have multiple lines. I'll use comma_separated_strings

mjambon replied:
No, it's that `"1,2",3` is 2 CSV cells holding `1,2` and `3`.

mjambon commented on libs/commons/Logs_.mli:
```diff
@@ -1,4 +1,4 @@
-(* Small helpers around the Logs library.
+(* Small wrapper around the Logs library.
```
It's not small...

aryx replied:
True ... Logs.ml is actually the same size ... 330 LOC.

mjambon commented on src/osemgrep/core/CLI_common.ml:
```diff
@@ -75,7 +75,8 @@ let setup_logging ~force_color ~level =
   Std_msg.setup ~highlight_setting:(if force_color then On else Auto) ();
-  Logs_.setup_logging ~level ();
+  (* coupling: See also calls to Logs_.setup() in Core_CLI.ml *)
+  Logs_.setup ~level ();
```
This looks like an appropriate spot to enable the `SEMGREP_LOG_` environment variables.

aryx replied:
yep will do

aryx committed:
adress iago and martin's comments

use SEMGREP_XXX and PYTEST_XXX env vars

test plan:
SEMGREP_LOG_SRCS=pcre make check

aryx committed:
cleanup changelog
